File: python/life.py
import random
import os
import time
from dotenv import load_dotenv

# Load .env from parent directory
load_dotenv(dotenv_path="../.env")

WIDTH = int(os.getenv("WIDTH", 40))
HEIGHT = int(os.getenv("HEIGHT", 20))
SPEED = float(os.getenv("SPEED", 0.15))
DENSITY = float(os.getenv("DENSITY", 0.75))


def create_grid():
    return [
        [1 if random.random() < DENSITY else 0 for _ in range(WIDTH)]
        for _ in range(HEIGHT)
    ]


def count_neighbors(grid, x, y):
    count = 0

    for dy in [-1, 0, 1]:
        for dx in [-1, 0, 1]:
            if dx == 0 and dy == 0:
                continue

            nx = x + dx
            ny = y + dy

            if 0 <= nx < WIDTH and 0 <= ny < HEIGHT:
                count += grid[ny][nx]

    return count


def next_generation(grid):
    return [
        [
            1
            if (grid[y][x] == 1 and count_neighbors(grid, x, y) in [2, 3])
               or (grid[y][x] == 0 and count_neighbors(grid, x, y) == 3)
            else 0
            for x in range(WIDTH)
        ]
        for y in range(HEIGHT)
    ]


def clear():
    os.system("cls" if os.name == "nt" else "clear")


def render(grid):
    clear()
    for row in grid:
        print("".join("█" if cell else " " for cell in row))


grid = create_grid()

while True:
    render(grid)
    grid = next_generation(grid)
    time.sleep(SPEED)

File: typescript/src/life.ts
import * as dotenv from "dotenv";

dotenv.config({ path: "../.env" });

type Grid = boolean[][];

const width = intEnv("WIDTH", 40);
const height = intEnv("HEIGHT", 20);
const speed = floatEnv("SPEED", 0.15);
const density = floatEnv("DENSITY", 0.25);

function intEnv(name: string, fallback: number): number {
    const value = process.env[name];
    const parsed = value ? Number.parseInt(value, 10) : NaN;
    return Number.isNaN(parsed) ? fallback : parsed;
}

function floatEnv(name: string, fallback: number): number {
    const value = process.env[name];
    const parsed = value ? Number.parseFloat(value) : NaN;
    return Number.isNaN(parsed) ? fallback : parsed;
}

function createGrid(): Grid {
    return Array.from({ length: height }, () =>
        Array.from({ length: width }, () => Math.random() < density)
    );
}

function isAlive(grid: Grid, x: number, y: number): boolean {
    return grid[y]?.[x] ?? false;
}

function countNeighbors(grid: Grid, x: number, y: number): number {
    let count = 0;

    for (let dy = -1; dy <= 1; dy++) {
        for (let dx = -1; dx <= 1; dx++) {
            if (dx === 0 && dy === 0) continue;

            if (isAlive(grid, x + dx, y + dy)) {
                count++;
            }
        }
    }

    return count;
}

function nextCell(grid: Grid, x: number, y: number): boolean {
    const alive = isAlive(grid, x, y);
    const neighbors = countNeighbors(grid, x, y);

    return (
        (alive && (neighbors === 2 || neighbors === 3)) ||
        (!alive && neighbors === 3)
    );
}

function nextGeneration(grid: Grid): Grid {
    return Array.from({ length: height }, (_, y) =>
        Array.from({ length: width }, (_, x) => nextCell(grid, x, y))
    );
}

function clearScreen(): void {
    process.stdout.write("\x1b[H\x1b[2J");
}

function render(grid: Grid): void {
    clearScreen();

    for (const row of grid) {
        console.log(row.map(cell => (cell ? "█" : " ")).join(""));
    }
}

let grid = createGrid();

setInterval(() => {
    render(grid);
    grid = nextGeneration(grid);
}, speed * 1000);